### PR TITLE
Support dmatrix construction from cupy array

### PR DESCRIPTION
**Pull request by RAMitchell**
A few things left to do:

- [x] Correctly construct labels/info
- [x] Add an integration test for training
- [x] Mention cupy in documentation

@trivialfis I removed the check for array interface version as I found I could get version 0 by installing cupy from conda or 2 by installing from pip, but not 1. Does this break anything for cudf?

### CODE REVIEW

**Comment by trivialfis** [comment]
No.  But would be nice the have it somewhere in the path.  I remember last time someone has a super old numba with cudf and caused a failure.

File: tests/python/testing.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # coding: utf-8
 from xgboost.compat import SKLEARN_INSTALLED, PANDAS_INSTALLED, DT_INSTALLED
-from xgboost.compat import CUDF_INSTALLED, DASK_INSTALLED
+from xgboost.compat import CUDF_INSTALLED, CUPY_INSTALLED, DASK_INSTALLED
```
Comment (trivialfis) [inline tests/python/testing.py]: Can we merge this after the release?  I would like to investigate lazy library loading.  These libraries are all heavy in size.

Comment (RAMitchell) [reply tests/python/testing.py]: Yep, I just measured it takes 0.6s to import cupy vs 1.73s for xgboost.

**Comment by RAMitchell** [comment]
@trivialfis I got this working without any imports. It might be possible to do the same with cudf, particularly if they fix the incomplete `__cuda_array_interface__` attribute.

**Comment by jakirkham** [comment]
> It might be possible to do the same with cudf, particularly if they fix the incomplete `__cuda_array_interface__` attribute.

@RAMitchell, could you please raise an issue on the cuDF repo documenting in which ways it is incomplete?

**Comment by RAMitchell** [comment]
@jakirkham it was my misunderstanding. I thought the mask attribute was supposed to be a `__cuda_array_interface__` , looking closer at the standard it should be an object that has the attribute `__cuda_array_interface__`.

File: python-package/xgboost/core.py [hunk 2]
```diff
@@ -862,7 +848,7 @@ def set_group(self, group):
         group : array like
             Group size of each group
         """
-        if _use_columnar_initializer(group):
+        if hasattr(group, "__cuda_array_interface__"):
```
Comment (trivialfis) [inline python-package/xgboost/core.py]: Does a cudf dataframe with single column have the array interface?  Also does a cudf series have array interface?  Can you be sure that the cudf test is using correct initializer?

File: src/c_api/c_api.cu [hunk 3]
```diff
@@ -7,14 +7,28 @@
...
   std::string json_str{c_json_strs};
   data::CudfAdapter adapter(json_str);
-  *out = new std::shared_ptr<DMatrix>(DMatrix::Create(&adapter, missing, 1));
+  *out =
+      new std::shared_ptr<DMatrix>(DMatrix::Create(&adapter, missing, nthread));
   API_END();
 }
+
+XGB_DLL int XGDMatrixCreateFromArrayInterface(char const* c_json_strs,
+                                                     bst_float missing,
```
Comment (trivialfis) [inline src/c_api/c_api.cu]: Wrong indent.

File: python-package/xgboost/core.py (hunk 2 above)
Comment (trivialfis) [reply python-package/xgboost/core.py]: > Does a cudf dataframe with single column have the array interface?

If it does, then we need to make sure the cudf DataFrame is not initialized with `cupy` initializer.

File: python-package/xgboost/core.py [hunk 4]
```diff
@@ -500,8 +461,10 @@ def __init__(self, data, label=None, weight=None, base_margin=None,
             self._init_from_npy2d(data, missing, nthread)
         elif isinstance(data, DataTable):
             self._init_from_dt(data, nthread)
-        elif _use_columnar_initializer(data):
-            self._init_from_columnar(data, missing)
+        elif hasattr(data, "__cuda_array_interface__"):
+            self._init_from_array_interface(data, missing, nthread)
+        elif isinstance(data, CUDF_DataFrame):
```
[the remark this answers is not in the document]
Comment (RAMitchell) [reply python-package/xgboost/core.py]: That sounds great if possible.

File: python-package/xgboost/core.py (hunk 2 above)
Comment (RAMitchell) [reply python-package/xgboost/core.py]: cudf with one column does not have array interface. I will modify it to accept this for metainfo.

**Comment by jakirkham** [comment]
> it was my misunderstanding. I thought the mask attribute was supposed to be a `__cuda_array_interface__` , looking closer at the standard it should be an object that has the attribute `__cuda_array_interface__`.

No worries. Just want to make sure when we are discovering issues they get raised so we can address them 🙂

**Pull request merged**